File: src/services/scaffolder.py
"""
Scaffolder Service - Cria estrutura de novos projetos.
"""
from pathlib import Path
from jinja2 import Template
from libs.prompt_templates import PromptTemplates
from libs.file_operations import FileOperations

class ScaffolderService:
    """Serviço de scaffolding de projetos."""
    
    def __init__(self, model_router, config):
        self.router = model_router
        self.config = config
        self.file_ops = FileOperations()
        self.templates_dir = Path(__file__).parent.parent / "templates"
    
    async def generate_structure(
        self,
        project_type: str,
        name: str,
        language: str,
        framework: str = None,
        database: str = None,
        include_auth: bool = False
    ) -> Path:
        """Gera estrutura completa do projeto."""
        # Template base
        template_name = self._select_template(project_type, language, framework)
        
        # Criar diretório
        project_path = Path.cwd() / name
        self.file_ops.create_directory(project_path)
        
        # Gerar README
        readme = await self._generate_readme({
            "name": name,
            "type": project_type,
            "language": language,
            "framework": framework,
            "database": database,
            "include_auth": include_auth
        })
        self.file_ops.write_file(project_path / "README.md", readme)
        
        # Estrutura básica dependendo do tipo
        if project_type == "api" and language == "python":
            await self._create_python_api(project_path, name, framework, database, include_auth)
        elif project_type == "webapp" and language == "typescript":
            await self._create_typescript_webapp(project_path, name, framework)
        
        return project_path
    
    def _select_template(self, project_type: str, language: str, framework: str) -> str:
        templates = {
            ("api", "python", "fastapi"): "api-fastapi",
            ("api", "typescript", "express"): "api-express",
            ("webapp", "typescript", "angular"): "webapp-angular",
        }
        return templates.get((project_type, language, framework), "api-fastapi")
    
    async def _generate_readme(self, config: dict) -> str:
        prompt = PromptTemplates.generate_readme(config)
        response = await self.router.complete(prompt, task_type="coding", max_tokens=2048)
        return response.response
    
    async def _create_python_api(self, path: Path, name: str, framework: str, database: str, auth: bool):
        """Cria estrutura de API Python."""
        # src/
        src = path / "src"
        self.file_ops.create_directory(src)
        
        # main.py
        main_content = f'''from fastapi import FastAPI

app = FastAPI(title="{name}")

@app.get("/")
def root():
    return {{"message": "{name} API is running"}}

@app.get("/health")
def health():
    return {{"status": "healthy"}}
'''
        self.file_ops.write_file(src / "main.py", main_content)
        
        # requirements.txt
        requirements = "fastapi>=0.109.0\nuvicorn[standard]>=0.27.0\n"
        if database:
            requirements += "sqlalchemy>=2.0.0\n"
        if auth:
            requirements += "python-jose[cryptography]>=3.3.0\npasslib[bcrypt]>=1.7.4\n"
        
        self.file_ops.write_file(path / "requirements.txt", requirements)
        
        # tests/
        self.file_ops.create_directory(path / "tests")
        
    async def _create_typescript_webapp(self, path: Path, name: str, framework: str):
        """Cria estrutura de webapp TypeScript."""
        # package.json
        package_json = f'''{{
  "name": "{name}",
  "version": "1.0.0",
  "scripts": {{
    "dev": "ng serve",
    "build": "ng build",
    "test": "ng test"
  }}
}}
'''
        self.file_ops.write_file(path / "package.json", package_json)

File: src/services/scaffolder.ts
import * as fs from 'fs';
import * as path from 'path';
import { ModelRouter } from '../core/modelRouter';
import { AppConfig } from '../types';
import { TemplateEngine, TemplateVars } from './templateEngine';

interface ScaffoldOptions {
  projectType: string;
  name: string;
  language: string;
  framework?: string;
  database?: string;
  includeAuth?: boolean;
}

export class ScaffolderService {
  private templateEngine: TemplateEngine;

  constructor(
    private router: ModelRouter,
    private config: AppConfig,
  ) {
    this.templateEngine = new TemplateEngine();
  }

  async generateStructure(opts: ScaffoldOptions): Promise<string> {
    const projectPath = path.join(process.cwd(), opts.name);
    fs.mkdirSync(projectPath, { recursive: true });

    const vars: TemplateVars = {
      projectName: opts.name,
      author: this.config.defaultAuthor,
      language: opts.language,
      framework: opts.framework,
      database: opts.database,
      includeAuth: opts.includeAuth,
      license: this.config.defaultLicense,
    };

    const templateKey = this.getTemplateKey(opts);

    if (templateKey && this.templateEngine.hasTemplate(templateKey)) {
      this.templateEngine.renderDirectory(templateKey, projectPath, vars);
    } else {
      this.fallbackScaffold(projectPath, opts);
    }

    // Generate README with LLM
    const readme = await this.generateReadme(opts);
    fs.writeFileSync(path.join(projectPath, 'README.md'), readme);

    return projectPath;
  }

  private getTemplateKey(opts: ScaffoldOptions): string | null {
    if (opts.projectType === 'api' && opts.language === 'python') return 'api-fastapi';
    if (opts.projectType === 'api' && opts.language === 'typescript') return 'api-express';
    if (opts.projectType === 'script' && opts.language === 'typescript') return 'script-typescript';
    if (opts.projectType === 'script') return 'script-typescript';
    return null;
  }

  private async generateReadme(opts: ScaffoldOptions): Promise<string> {
    const prompt = `Generate a README.md for project "${opts.name}" (${opts.projectType}, ${opts.language}${opts.framework ? ', ' + opts.framework : ''}). Include: title, description, install, usage, env vars, license. Be concise. Return only markdown.`;
    const res = await this.router.complete(prompt, 'coding', 512);
    return res.response;
  }

  private fallbackScaffold(projectPath: string, opts: ScaffoldOptions): void {
    if (opts.projectType === 'api' && opts.language === 'python') {
      this.createPythonApi(projectPath, opts);
    } else if (opts.projectType === 'api' && opts.language === 'typescript') {
      this.createExpressApi(projectPath, opts);
    } else if (opts.projectType === 'webapp') {
      this.createWebapp(projectPath, opts);
    } else if (opts.projectType === 'script') {
      this.createScript(projectPath, opts);
    }
  }

  private createPythonApi(projectPath: string, opts: ScaffoldOptions): void {
    const src = path.join(projectPath, 'src');
    fs.mkdirSync(src, { recursive: true });
    fs.mkdirSync(path.join(projectPath, 'tests'), { recursive: true });

    fs.writeFileSync(
      path.join(src, 'main.py'),
      `from fastapi import FastAPI\n\napp = FastAPI(title="${opts.name}")\n\n@app.get("/health")\ndef health():\n    return {"status": "healthy"}\n`,
    );

    let requirements = 'fastapi>=0.109.0\nuvicorn[standard]>=0.27.0\n';
    if (opts.database) requirements += 'sqlalchemy>=2.0.0\n';
    if (opts.includeAuth) requirements += 'python-jose[cryptography]>=3.3.0\n';
    fs.writeFileSync(path.join(projectPath, 'requirements.txt'), requirements);

    this.writeDockerfile(projectPath, 'python');
  }

  private createExpressApi(projectPath: string, opts: ScaffoldOptions): void {
    const src = path.join(projectPath, 'src');
    fs.mkdirSync(src, { recursive: true });
    fs.mkdirSync(path.join(projectPath, 'tests'), { recursive: true });

    fs.writeFileSync(
      path.join(src, 'app.ts'),
      `import express from 'express';\n\nconst app = express();\napp.use(express.json());\n\napp.get('/health', (_req, res) => res.json({ status: 'healthy' }));\n\nexport default app;\n`,
    );

    const pkg = {
      name: opts.name,
      version: '1.0.0',
      scripts: { dev: 'ts-node src/app.ts', build: 'tsc', test: 'jest' },
      dependencies: { express: '^4.18.0' },
      devDependencies: { typescript: '^5.5.0', '@types/express': '^4.17.0', 'ts-node': '^10.9.0' },
    };
    fs.writeFileSync(path.join(projectPath, 'package.json'), JSON.stringify(pkg, null, 2));

    this.writeDockerfile(projectPath, 'node');
  }

  private createWebapp(projectPath: string, opts: ScaffoldOptions): void {
    fs.mkdirSync(path.join(projectPath, 'src'), { recursive: true });
    const pkg = {
      name: opts.name,
      version: '1.0.0',
      scripts: { dev: 'ng serve', build: 'ng build', test: 'ng test' },
    };
    fs.writeFileSync(path.join(projectPath, 'package.json'), JSON.stringify(pkg, null, 2));
  }

  private createScript(projectPath: string, opts: ScaffoldOptions): void {
    fs.mkdirSync(path.join(projectPath, 'src'), { recursive: true });
    fs.writeFileSync(
      path.join(projectPath, 'src', 'index.ts'),
      `#!/usr/bin/env ts-node\n// ${opts.name}\n\nasync function main(): Promise<void> {\n  console.log('${opts.name} running...');\n}\n\nmain().catch(console.error);\n`,
    );
  }

  private writeDockerfile(projectPath: string, runtime: 'python' | 'node'): void {
    const content =
      runtime === 'python'
        ? `FROM python:3.11-slim\nWORKDIR /app\nCOPY requirements.txt .\nRUN pip install -r requirements.txt\nCOPY . .\nCMD ["uvicorn", "src.main:app", "--host", "0.0.0.0", "--port", "8000"]\n`
        : `FROM node:20-alpine\nWORKDIR /app\nCOPY package*.json .\nRUN npm ci\nCOPY . .\nRUN npm run build\nCMD ["node", "dist/app.js"]\n`;
    fs.writeFileSync(path.join(projectPath, 'Dockerfile'), content);
  }
}
